Provide a getter for webViewConfiguration on _WKWebExtensionContext.

xeenon opened the pull request:
#### 82de229ab5a54e1441857befa56fa9cab65b1e8e
<pre>
Provide a getter for webViewConfiguration on _WKWebExtensionContext.
<a href="https://webkit.org/b/263533">https://webkit.org/b/263533</a>
rdar://problem/117359849

Reviewed by Brian Weinstein.

Add API for obtaining correct WKWebViewConfiguration for web extension content in tabs.
Flesh out the remaining web view configuration properties needed, and provide configurations
for distinct web view purposes with different process names in manifest v3.

To make sure the app uses the right configuration, make extension URL load fail if they are not
using the right configuration, and prevent extension configured views from loading other URLs.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]): Copy _requiredWebExtensionBaseURL.
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]): Added.
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext webViewConfiguration]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensionContextForURL:]): Added.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::isUnsupportedWebExtensionNavigation): Added.
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction): Call isUnsupportedWebExtensionNavigation.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::popupWebView): Pass a WebViewPurpose for the configuration.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::webProcessDisplayName): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::relatedWebView): Return nil for v3 and later. Simplify by looking at all known pages that
have the extension URL loaded, so it catches tab pages too.
(WebKit::WebExtensionContext::processDisplayName): Added.
(WebKit::WebExtensionContext::corsDisablingPatterns): Added.
(WebKit::WebExtensionContext::webViewConfiguration): Expanded and added WebViewPurpose.
(WebKit::WebExtensionContext::loadBackgroundWebView): Pass a WebViewPurpose for the configuration.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Return an error when loading an extension URL in the main frame
of a improperly configured web view.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::supportsManifestVersion): Added an ASSERT to catch a mistake that I keep making where passing 2
is invalid since that will be true for all extensions.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: <a href="https://commits.webkit.org/269669@main">https://commits.webkit.org/269669@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b9bd80c3811c7bce704c40b730a4bf0f471be67

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/23214 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/1303 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/24334 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/25192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/21452 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/23483 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/2818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23752 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/25192 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/23454 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/895 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/20118 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25977 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/718 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21003 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/27166 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/21169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/21266 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/25041 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/708 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/18472 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/645 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/1119 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2950 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/920 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -1716,45 +1717,96 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
+    for (auto& requestedMatchPattern : requestedMatchPatterns)
+        [patterns addObjectsFromArray:requestedMatchPattern->expandedStrings()];
+
+    // Include manifest optional permission origins here, these should be dynamically added when the are granted
```
nit: typo
> // ... when the are granted

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtension.h:
```diff
@@ -147,7 +147,7 @@ class WebExtension : public API::ObjectImpl<API::Object::Type::WebExtension>, pu
 
     double manifestVersion();
-    bool supportsManifestVersion(double version) { return manifestVersion() >= version; }
+    bool supportsManifestVersion(double version) { ASSERT(version > 2); return manifestVersion() >= version; }
```
Is there no reason to call this with `supportsManifestVersion(2)`?

xeenon replied:
I mention this in the change log, but passing 2 will be true for every extension.

xeenon committed:
Provide a getter for webViewConfiguration on _WKWebExtensionContext.
https://webkit.org/b/263533
rdar://problem/117359849

Reviewed by Brian Weinstein.

Add API for obtaining correct WKWebViewConfiguration for web extension content in tabs.
Flesh out the remaining web view configuration properties needed, and provide configurations
for distinct web view purposes with different process names in manifest v3.

To make sure the app uses the right configuration, make extension URL load fail if they are not
using the right configuration, and prevent extension configured views from loading other URLs.

* Source/WebCore/en.lproj/Localizable.strings: Updated.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfiguration.mm:
(-[WKWebViewConfiguration copyWithZone:]): Copy _requiredWebExtensionBaseURL.
(-[WKWebViewConfiguration _requiredWebExtensionBaseURL]): Added.
(-[WKWebViewConfiguration _setRequiredWebExtensionBaseURL:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewConfigurationPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext webViewConfiguration]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController extensionContextForURL:]): Added.
* Source/WebKit/UIProcess/Cocoa/NavigationState.mm:
(WebKit::isUnsupportedWebExtensionNavigation): Added.
(WebKit::NavigationState::NavigationClient::decidePolicyForNavigationAction): Call isUnsupportedWebExtensionNavigation.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::popupWebView): Pass a WebViewPurpose for the configuration.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionCocoa.mm:
(WebKit::WebExtension::webProcessDisplayName): Deleted.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::relatedWebView): Return nil for v3 and later. Simplify by looking at all known pages that
have the extension URL loaded, so it catches tab pages too.
(WebKit::WebExtensionContext::processDisplayName): Added.
(WebKit::WebExtensionContext::corsDisablingPatterns): Added.
(WebKit::WebExtensionContext::webViewConfiguration): Expanded and added WebViewPurpose.
(WebKit::WebExtensionContext::loadBackgroundWebView): Pass a WebViewPurpose for the configuration.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionURLSchemeHandlerCocoa.mm:
(WebKit::WebExtensionURLSchemeHandler::platformStartTask): Return an error when loading an extension URL in the main frame
of a improperly configured web view.
* Source/WebKit/UIProcess/Extensions/WebExtension.h:
(WebKit::WebExtension::supportsManifestVersion): Added an ASSERT to catch a mistake that I keep making where passing 2
is invalid since that will be true for all extensions.
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: https://commits.webkit.org/269669@main
